README.md
Bukanir
=======

Bukanir streams movies and TV shows from bittorrent magnet links

Features
--------

* Watch movies and TV shows on your Desktop or Android devices
* Embedded Tor for privacy
* Many languages to choose for subtitles
* Ads free and free as in Freedom

License
-------

Bukanir is free/libre software released under the terms of the GNU GPL license,
see the `COPYING' file for details.


CHANGELOG
v2.5  Mar 13, 2022:

    * Tor is embedded in the binary
    * Bug fixes

v2.4  Mar 9, 2018:

    * Added support for tor socks5 proxy and .onion links
    * Desktop app now supports translations for UI
    * Added search by cast or director
    * Updated ijkplayer (FFmpeg v3.4) and libtorrent-rasterbar (v1.0.11)

v2.3  Oct 20, 2016:

    * Desktop app is rewritten in Go (Qt5)
    * Added eztv search provider
    * Search by genre
    * Updated ijkplayer (FFmpeg v3.1) and libtorrent-rasterbar (v1.0.10)

v2.2  Jun 22, 2016:

    * Speedup loading of videos

v2.1  Jun 6, 2016:

    * Torrent2http is now bundled inside bukanir lib
    * Fixed crash on arm64

v2.0  Oct 13, 2015:

    * Fixed tpb domains
    * Add support for arm64
    * Update libtorrent to 1.1.0

v1.9  Oct 13, 2015:

    * Seeking in videos now works, but it must be enabled in settings, player startup time will be longer though
    * Vitamio player is replaced with ijkplayer which currently uses FFmpeg version 2.8
    * Trailers and favorites are added, also, genre and quality for HD Movies/HD Episodes are now shown
    * Fetching metadata from themoviedb.org is much faster
    * Subtitles, subtitle delay and video aspect can be changed during playback
    * Added subscene.com subtitle provider
    * Subtitle matching is now more precise
    * Backend is compiled with Go version 1.5 and libtorrent 1.0.6

v1.8  May 21, 2015:

    * Implement rate limit for tmdb
    * Add new tpb domains

v1.7  May 17, 2015:

    * Fix for podnapisi.net old api

v1.6  Apr 20, 2015:

    * Add opensubtitles support
    * Add search suggestions

v1.5  Apr 15, 2015:

    * Add tv episodes
    * Fix metadata and subtitles fetching

v1.4  Apr 7, 2015:

    * Fix metadata fetching
    * Fix for external sd card location

v1.3  Mar 19, 2015:

    * Piratebay is back

v1.2  Sep 12, 2014:

    * Greatly speed up downloading metadata
    * Fix subtitles download

v1.1  Aug 10, 2014:

    * Fix issue with tpb
    * Compiled against libtorrent-1.0
    * Bug fixes

v1.0  Jun 17, 2014:

    * First public release


lib/subs.go
package bukanir

import (
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"log"
	"net/url"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"

	"github.com/PuerkitoBio/goquery"
	"github.com/oz/osdb"
	"github.com/xrash/smetrics"
)

// language type
type language struct {
	Name      string
	ISO_639_2 string
	ID        string
}

// Minimum score required
var scoreRequired float64 = 0.65

// Supported languages
var languages = []language{
	{"Albanian", "alb", "29"},
	{"Arabic", "ara", "12"},
	{"Belarus", "bel", "50"},
	{"Bengali", "ben", "59"},
	{"Bosnian", "bos", "10"},
	{"Bulgarian", "bul", "33"},
	{"Catalan", "cat", "53"},
	{"Chinese", "zho", "17"},
	{"Croatian", "hrv", "38"},
	{"Czech", "ces", "7"},
	{"Danish", "dan", "24"},
	{"Dutch", "dut", "23"},
	{"English", "eng", "2"},
	{"Estonian", "est", "20"},
	{"Finnish", "fin", "31"},
	{"French", "fra", "8"},
	{"German", "ger", "5"},
	{"Greek", "gre", "16"},
	{"Hebrew", "heb", "22"},
	{"Hindi", "hin", "42"},
	{"Hungarian", "hun", "15"},
	{"Icelandic", "isl", "6"},
	{"Indonesian", "ind", "54"},
	{"Irish", "gle", "49"},
	{"Italian", "ita", "9"},
	{"Japanese", "jpn", "11"},
	{"Kazakh", "kaz", "58"},
	{"Korean", "kor", "4"},
	{"Latvian", "lav", "21"},
	{"Lithuanian", "lit", "19"},
	{"Macedonian", "mkd", "35"},
	{"Malay", "msa", "55"},
	{"Norwegian", "nor", "3"},
	{"Polish", "pol", "26"},
	{"Portuguese", "por", "32"},
	{"Romanian", "ron", "13"},
	{"Russian", "rus", "27"},
	{"Serbian", "srp", "36"},
	{"Sinhala", "sin", "56"},
	{"Slovak", "slk", "37"},
	{"Slovenian", "slv", "1"},
	{"Spanish", "spa", "28"},
	{"Swedish", "swe", "25"},
	{"Thai", "tha", "44"},
	{"Turkish", "tur", "30"},
	{"Ukrainian", "ukr", "46"},
	{"Vietnamese", "vie", "51"},
}

// podnapisi search for subtitles on podnapisi.net
func podnapisi(movie string, year string, torrentRelease string, lang string, category int, season int, episode int) {
	defer wgs.Done()
	defer func() {
		if r := recover(); r != nil {
			log.Print("SUB: Recovered in podnapisi")
		}
	}()

	type Page struct {
		Current int `xml:"current"`
		Count   int `xml:"count"`
		Results int `xml:"results"`
	}

	type Subtitle struct {
		XMLName    xml.Name `xml:"subtitle"`
		Id         int      `xml:"id"`
		Pid        string   `xml:"pid"`
		Title      string   `xml:"title"`
		Year       string   `xml:"year"`
		Url        string   `xml:"url"`
		Release    string   `xml:"release"`
		TvSeason   int      `xml:"tvSeason"`
		TvEpisode  int      `xml:"tvEpisode"`
		Language   string   `xml:"language"`
		LanguageID int      `xml:"languageId"`
	}

	type Data struct {
		XMLName      xml.Name   `xml:"results"`
		Pagination   Page       `xml:"pagination"`
		SubtitleList []Subtitle `xml:"subtitle"`
	}

	l := getLanguage(lang)

	searchUrl := "http://podnapisi.net/subtitles/search/old?sXML=1&sK=%s&sY=%s&sJ=%s"
	uri := fmt.Sprintf(searchUrl, url.QueryEscape(movie), year, l.ID)

	if category == CategoryTV || category == CategoryHDTV {
		if season != 0 {
			uri = uri + fmt.Sprintf("&sTS=%d&sTE=%d", season, episode)
		}
	}

	if verbose {
		log.Printf("SUB: GET %s\n", uri)
	}

	res, err := getResponse(uri)
	if err != nil {
		log.Printf("ERROR: getResponse %s\n", err.Error())
		return
	}

	defer res.Body.Close()
	body, err := ioutil.ReadAll(res.Body)
	if err != nil {
		log.Printf("ERROR: ReadAll: %s\n", err.Error())
		return
	}

	var r Data
	err = xml.Unmarshal(body, &r)
	if err != nil {
		log.Printf("ERROR: Unmarshal: %s\n", err.Error())
		return
	}

	if len(r.SubtitleList) == 0 {
		return
	}

	for _, s := range r.SubtitleList {
		rel := strings.Fields(s.Release)
		var subtitleRelease string
		if len(rel) > 0 {
			subtitleRelease = rel[0]
		} else {
			continue
		}

		lid, _ := strconv.Atoi(l.ID)
		if s.LanguageID != lid {
			continue
		}

		score := getSubScore(torrentRelease, subtitleRelease)
		if score < scoreRequired {
			continue
		}

		if category == CategoryTV || category == CategoryHDTV {
			if season != s.TvSeason || episode != s.TvEpisode {
				continue
			}
		}

		downloadLink := fmt.Sprintf("http://podnapisi.net/subtitles/%s/download", s.Pid)

		s := TSubtitle{strconv.Itoa(s.Id), s.Title, s.Year, subtitleRelease, downloadLink, score}
		subtitles = append(subtitles, s)
	}
}

// opensubtitles search for subtitles on opensubtitles.org
func opensubtitles(movie string, imdbId string, year string, torrentRelease string, lang string, category int, season int, episode int) {
	defer wgs.Done()
	defer func() {
		if r := recover(); r != nil {
			log.Print("SUB: Recovered in opensubtitles")
		}
	}()

	o, err := osdb.NewClient()
	if err != nil {
		log.Printf("ERROR: NewClient: %s\n", err.Error())
	}

	err = o.LogIn(opensubsUser, opensubsPassword, "en")
	if err != nil {
		log.Printf("ERROR: LogIn: %s\n", err.Error())
	}

	o.UserAgent = opensubsUserAgent

	params := []interface{}{
		o.Token,
		[]map[string]string{},
	}

	l := getLanguage(lang)
	var listArgs []map[string]string
	listArgs = append(listArgs, map[string]string{"imdbid": imdbId})
	listArgs = append(listArgs, map[string]string{"sublanguageid": l.ISO_639_2})

	if category == CategoryTV || category == CategoryHDTV {
		if season != 0 {
			listArgs = append(listArgs, map[string]string{"season": strconv.Itoa(season)})
			listArgs = append(listArgs, map[string]string{"episode": strconv.Itoa(episode)})
		}
	}

	if verbose {
		log.Printf("SUB: XML-RPC http://opensubtitles.org %+v\n", listArgs)
	}

	params[1] = listArgs
	subs, err := o.SearchSubtitles(&params)
	if err != nil {
		log.Printf("ERROR: SearchSubtitles: %v\n", err.Error())
		return
	}

	if len(subs) == 0 {
		return
	}

	for _, sub := range subs {
		if sub.SubLanguageID != l.ISO_639_2 {
			continue
		}

		if !isSubValid(sub.SubFormat) {
			continue
		}

		subSumCD, _ := strconv.Atoi(sub.SubSumCD)
		if subSumCD > 1 {
			continue
		}

		score := getSubScore(torrentRelease, sub.MovieReleaseName)
		if score < scoreRequired {
			continue
		}

		if category == CategoryTV || category == CategoryHDTV {
			subSeason, _ := strconv.Atoi(sub.SeriesSeason)
			subEpisode, _ := strconv.Atoi(sub.SeriesEpisode)
			if season != subSeason || episode != subEpisode {
				continue
			}
		}

		s := TSubtitle{sub.IDSubtitleFile, sub.MovieName, sub.MovieYear, sub.MovieReleaseName, sub.ZipDownloadLink, score}
		subtitles = append(subtitles, s)
	}

}

// subscene search for subtitles on subscene.com
func subscene(movie string, year string, torrentRelease string, lang string, category int, season int, episode int) {
	defer wgs.Done()
	defer func() {
		if r := recover(); r != nil {
			log.Print("SUB: Recovered in subscene")
		}
	}()

	searchUrl := "http://subscene.com/subtitles/release?q=%s"

	uri := fmt.Sprintf(searchUrl, url.QueryEscape(torrentRelease))

	if verbose {
		log.Printf("SUB: GET %s\n", uri)
	}

	doc, err := getDocument(uri)
	if err != nil {
		log.Printf("ERROR: getDocument %s", err.Error())
		return
	}

	tds := doc.Find(`td.a1`)

	if tds.Length() == 0 {
		return
	}

	subs := make([]TSubtitle, 0)

	var w sync.WaitGroup
	parseHTML := func(i int, td *goquery.Selection) {
		defer w.Done()
		defer func() {
			if r := recover(); r != nil {
				log.Print("SUB: Recovered in parseHTML")
			}
		}()

		link := td.Find(`a`).First()
		href, _ := link.Attr("href")
		id := filepath.Base(href)

		subtitleLang := td.Find(`span`).First().Text()
		subtitleRelease := td.Find(`span`).Next().Text()
		subtitleYear := getYear(subtitleRelease)

		subtitleLang = strings.TrimSpace(subtitleLang)
		subtitleRelease = strings.TrimSpace(subtitleRelease)
		subtitleYear = strings.TrimSpace(subtitleYear)

		subtitleTitle := getTitle(subtitleRelease)

		if strings.ToLower(subtitleTitle) != strings.ToLower(movie) {
			return
		}

		if strings.ToLower(subtitleLang) != strings.ToLower(lang) {
			return
		}

		if subtitleYear != "" && subtitleYear != year {
			return
		}

		score := getSubScore(torrentRelease, subtitleRelease)
		if score < scoreRequired {
			return
		}

		if category == CategoryTV || category == CategoryHDTV {
			rs, _ := strconv.Atoi(getSeason(subtitleRelease))
			re, _ := strconv.Atoi(getEpisode(subtitleRelease))
			if season != rs || episode != re {
				return
			}
		}

		s := TSubtitle{id, subtitleTitle, subtitleYear, subtitleRelease, href, score}
		subs = append(subs, s)
	}

	w.Add(tds.Length())
	tds.Each(func(i int, s *goquery.Selection) {
		go parseHTML(i, s)
	})
	w.Wait()

	if len(subs) == 0 {
		return
	}

	sort.Sort(ByScore(subs))
	s := subs[0]

	d, err := getDocument("http://subscene.com" + s.DownloadLink)
	if err != nil {
		log.Printf("ERROR: getDocument %s\n", err.Error())
		return
	}

	downloadHref := d.Find("#downloadButton")
	downloadLink, _ := downloadHref.Attr("href")

	sub := TSubtitle{"0", s.Title, s.Year, s.Release, "http://subscene.com" + downloadLink, s.Score}
	subtitles = append(subtitles, sub)
}

// getSubScore returns subtitle score
func getSubScore(torrentRelease string, subtitleRelease string) float64 {
	for _, char := range []string{".", "-", "_"} {
		torrentRelease = strings.Replace(torrentRelease, char, " ", -1)
		subtitleRelease = strings.Replace(subtitleRelease, char, " ", -1)
	}
	return smetrics.Jaro(torrentRelease, subtitleRelease)
}

// isSubValid checks if subtitle is valid
func isSubValid(format string) bool {
	formats := []string{"srt", "ass", "ssa"}
	if runtime.GOOS != "android" {
		formats = append(formats, "sub")
	}
	for _, f := range formats {
		if f == format {
			return true
		}
	}
	return false
}
